lib/token_store.rb

```diff
@@ -1,17 +1,44 @@
 class TokenStore
+  cattr_reader :token_caches
+
   @token_caches = {} # Hash of Memory/Dalli Store Caches, Keyed by namespace
 
+  module KeyValuehelpers
+    def delete_all_for_user(userid)
+      Array(read("tokens_for_#{userid}")).each do |token|
+        delete(token)
+      end
+
+      delete("tokens_for_#{userid}")
+    end
+
+    def create_user_token(token, data, options)
+      write(token, data, options)
+    ensure
+      if data[:userid]
+        user_tokens_cache_key = "tokens_for_#{data[:userid]}"
+        user_tokens_cache = read(user_tokens_cache_key) || []
+        user_tokens_cache << token
```

Comment:
This is really tricky in a multi-threaded environment particularly with memcached. As two competing threads come in, they can't step on each other.

This was sort of what I was hinting at in the original proposal (https://github.com/ManageIQ/manageiq/issues/20378#issue-663333118 - see the cross out section on memory_store), where I mentioned for memached we might need to use [CAS](https://www.tutorialspoint.com/memcached/memcached_cas.htm).

These particular "create an extra index in the memory store" was what we were trying to avoid by storing all of the records in the sessions table regardless of the store type.  I'm not against encoding an index in each one instead, but I think each one has nuances.

Reply:
if we are using multiple memcached instances, then each server will have a unique list of session ids
and has a low likelihood of stepping on another server's toes. Also, every server can not clear the sessions on other memcached servers. Maybe the use cases for problems are simpler in terms of the session store and more complicated in terms of broadcasting to delete all of a user's sessions.

Yes. A single server could have multiple sessions coming live at the same time. Maybe via the api and rails at the same time? I am having trouble coming up with a use case that doesn't seem far fetched.

Before we go super complex on distributed memcached servers, can we entertain using a single session store? Just like we use a single data store (postgres).